File: Python/Fraction to Recurring Decimal.py
class Solution:
    # @param {integer} numerator
    # @param {integer} denominator
    # @return {string}
    def fractionToDecimal(self, numerator, denominator):
        sig = numerator < 0 and -1 or 1
        sig = denominator < 0 and sig * -1 or sig
        numerator = abs(numerator)
        denominator = abs(denominator)
        mystr = ('%s' % (numerator // denominator)).split()
        if numerator and sig == -1:
            mystr.insert(0, '-')
        numerator %= denominator
        if numerator:
            mystr.append('.')
        hashmap = {}
        pt = len(mystr)
        while numerator and not hashmap.get(numerator):
            hashmap[numerator] = pt
            numerator *= 10
            mystr.append(str(numerator // denominator))
            numerator %= denominator
            pt += 1
        if numerator:
            mystr.insert(hashmap[numerator], '(')
            mystr.append(')')
        return "".join(mystr)

File: java/Fraction_to_Recurring_Decimal.java
// HashMap, Math
// Using a hashmap to record the remainder at each point.
// When we find the current remainder exits in the hashmap, we top the loop, and give out the result;
// We should use long to avoid the overflow cases of int!
// Time is TODO! space is TODO!

public class Solution {
    public String fractionToDecimal(int numerator, int denominator) {
        StringBuilder sb = new StringBuilder();
        HashMap<Long, Integer> mods = new HashMap<Long, Integer>();
        long a = Long.valueOf(numerator);
        long b = Long.valueOf(denominator);
        boolean neg = false;
        if (a < 0 && b > 0 || a > 0 && b < 0)
            neg = true;
        a = Math.abs(a);
        b = Math.abs(b);
        sb.append(a/b);
        long n = a%b;
        if (n == 0) {
            if (neg) sb.insert(0,'-');
            return sb.toString();
        }
        StringBuilder sbb = new StringBuilder();
        sb.append('.');
        while (n != 0) {
            if (mods.containsKey(n)) {
                int loc = mods.get(n);
                sbb.insert(loc,'(');
                sbb.append(')');
                break;
            }
            mods.put(n,sbb.length());
            n *= 10;
            while (n/b == 0) {
                sbb.append('0');
                mods.put(n,sbb.length());
                n *= 10;
            }
            long val = n/b;
            n = n%b;
            sbb.append(val);
        }
        sb.append(sbb);
        if (neg) sb.insert(0,'-');
        return sb.toString();
    }
}
